### PR TITLE
Script Editor to Ha Form

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

change Over Sequence editor in scripts to ha-form

 [ ] change first form to ha form

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Review by zsarnett — COMMENTED** [review]
I can't figure out why the schema for the mode and entity id are being continuously added everytime render happens...

image

File: src/components/ha-selector/ha-selector-icon.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  @property() public label?: string;
+
+  @property({ type: Boolean, reflect: true }) public disabled = false;
+
+  protected render() {
+    return html`
+      <ha-icon-picker
+        .label=${this.hass.localize("ui.panel.config.script.editor.icon")}
+        .name=${"icon"}
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-icon.ts:24]:
```suggestion
```

File: src/components/ha-selector/ha-selector-icon.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  @property() public value?: string;
+
+  @property() public label?: string;
+
+  @property({ type: Boolean, reflect: true }) public disabled = false;
+
+  protected render() {
+    return html`
+      <ha-icon-picker
+        .label=${this.hass.localize("ui.panel.config.script.editor.icon")}
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-icon.ts:23]:
```suggestion
        .label=${this.label}
```

File: src/components/ha-selector/ha-selector-select.ts [hunk 3]
```diff
@@ -25,15 +27,17 @@ export class HaSelectSelector extends LitElement {
...
       @selected=${this._valueChanged}
     >
-      ${this.selector.select.options.map(
-        (item: string) => html`
-          <mwc-list-item .value=${item}>${item}</mwc-list-item>
-        `
-      )}
+      ${this.selector.select.options.map((item: string | SelectOption) => {
+        const value = typeof item === "object" ? item.value : item;
+        const label = typeof item === "object" ? item.label : item;
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-select.ts:37]: Or should we do it like `ha-form`? 🤔 

<https://github.com/home-assistant/frontend/blob/dev/src/components/ha-form/ha-form-multi_select.ts#L24-L30>

Comment (bramkragten) [reply src/components/ha-selector/ha-selector-select.ts:37]: I like the object better, but not sure how well it translates to the backend

File: src/panels/config/script/ha-script-editor.ts [hunk 4]
```diff
@@ -52,12 +55,28 @@ import { haStyle } from "../../../resources/styles";
...
 import "../ha-config-section";
 import { configSections } from "../ha-panel-config";
 import "./blueprint-script-editor";
 
+const BASE_GUI_SCHEMA: HaFormSelector[] = [
+  {
+    name: "alias",
+    selector: {
+      text: {
+        type: "text",
```
Comment (bramkragten) [inline src/panels/config/script/ha-script-editor.ts:68]: This is the default, not strictly needed

File: src/panels/config/script/ha-script-editor.ts [hunk 5]
```diff
@@ -83,7 +102,75 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+  private _schema?: HaFormSelector[];
+
+  protected willUpdate(): void {
+    if (this._mode !== "gui" || !this._config) {
+      return;
+    }
+
+    const schema = BASE_GUI_SCHEMA;
+    if (!this.scriptEntityId) {
+      schema!.push({
```
Comment (bramkragten) [inline src/panels/config/script/ha-script-editor.ts:114]: You keep pushing this to the base schema every update, you should not update the base schema

File: src/panels/config/script/ha-script-editor.ts [hunk 6]
```diff
@@ -83,7 +102,75 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
 
   @query("ha-yaml-editor", true) private _editor?: HaYamlEditor;
 
+  private _schema?: HaFormSelector[];
+
+  protected willUpdate(): void {
```
Comment (bramkragten) [inline src/panels/config/script/ha-script-editor.ts:107]: Make this a memoized function

File: src/components/ha-form/ha-form.ts [hunk 7]
```diff
@@ -85,14 +91,14 @@ export class HaForm extends LitElement implements HaFormElement {
                   .hass=${this.hass}
                   .selector=${item.selector}
                   .value=${getValue(this.data, item)}
-                  .label=${this._computeLabel(item)}
-                  .disabled=${this.disabled}
```
Comment (balloob) [inline src/components/ha-form/ha-form.ts:89]: Why remove disabled?

File: src/components/ha-selector/ha-selector-icon.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  @property({ type: Boolean, reflect: true }) public disabled = false;
+
+  protected render() {
+    return html`
+      <ha-icon-picker
+        .label=${this.label}
+        .value=${this.value}
+        @value-changed=${this._valueChanged}
+      >
+      </ha-icon-picker>
```
Comment (balloob) [inline src/components/ha-selector/ha-selector-icon.ts:27]:
```suggestion
      ></ha-icon-picker>
```

File: src/panels/config/script/ha-script-editor.ts [hunk 9]
```diff
@@ -83,7 +102,78 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
 
   @query("ha-yaml-editor", true) private _editor?: HaYamlEditor;
 
+  private _schema = memoizeOne(
+    (hasID: boolean, useBluePrint?: boolean, currentMode?: string) => {
+      const schema: HaFormSchema[] = [];
```
Comment (balloob) [inline src/panels/config/script/ha-script-editor.ts:107]:
```suggestion
      const schema = [...BASE_GUI_SCHEMA];
```

File: src/panels/config/script/ha-script-editor.ts [hunk 10]
```diff
@@ -83,7 +102,78 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+          selector: {
+            text: {
+              type: "number",
+            },
+          },
+        });
+      }
+
+      schema.unshift(...BASE_GUI_SCHEMA);
+
```
Comment (balloob) [inline src/panels/config/script/ha-script-editor.ts:150]:
```suggestion
```

File: src/panels/config/script/ha-script-editor.ts (hunk 9 above)
Comment (balloob) [reply src/panels/config/script/ha-script-editor.ts:107]: Why even have a BASE_GUI_SCHEMA? You can just define the fields in here. The method is memoized

File: src/components/ha-form/ha-form.ts (hunk 7 above)
Comment (zsarnett) [reply src/components/ha-form/ha-form.ts:89]: bad rebase i think

File: src/panels/config/script/ha-script-editor.ts [hunk 11]
```diff
@@ -83,7 +85,91 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+        {
+          name: "icon",
+          selector: {
+            icon: {},
+          },
+        },
+      ];
+
+      if (!hasID) {
+        schema!.push({
```
Comment (bramkragten) [inline src/panels/config/script/ha-script-editor.ts:108]:
```suggestion
        schema.push({
```

File: src/panels/config/script/ha-script-editor.ts [hunk 12]
```diff
@@ -83,7 +85,91 @@ export class HaScriptEditor extends KeyboardShortcutMixin(LitElement) {
...
+                `,
+                value: mode,
+              })),
+            },
+          },
+        });
+      }
+
+      if (currentMode && MODES_MAX.includes(currentMode)) {
+        schema!.push({
```
Comment (bramkragten) [inline src/panels/config/script/ha-script-editor.ts:137]:
```suggestion
        schema.push({
```

**Review by bramkragten — APPROVED** [review]

**Pull request merged**